### PR TITLE
Dzelge webapi

### PR DESCRIPTION
**Pull request by dzelge**
When accepting this PR version 0.1.0 will be merged into the master. It will contain:

- The Core Python GeoDBClient 
- Sphinx docs
- Example notebooks
- A conda build recipe
- A Dockerfile

### CODE REVIEW

File: xcube_geodb/auth0/client_grant.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "client_id": "d479nNfAZJh3Ni0qvGtL0R5DLRZqNVno",
+  "audience": "geodb",
```
Comment (forman) [inline xcube_geodb/auth0/client_grant.json]: `xcube_geodb`

File: xcube_geodb/cli/get_by_bbox.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+@click.command(name='get_by_bbox')
+@click.option('--dataset', '-d', 'dataset', metavar='DATASET', help="Comma-separated list of bbox.")
+@click.option('--bbox', '-b', 'bbox', metavar='BBOX', help="Comma-separated list of bbox.")
+@click.option('--limit', '-l', 'limit', metavar='LIMIT', help="Limit or rows to be returned")
+@click.option('--offset', '-o', 'offset', metavar='OFFSET', help="Starting point")
+@click.option('--bbox-mode', '-m', 'bbox-mode', metavar='BBOX_MODE', help="Starting point")
+@click.option('--bbox-crs', '-c', 'bbox-crs', metavar='BBOX_CRS', help="Starting point")
+def get_by_bbox(dataset, bbox, limit, offset, bbox_mode, bbox_crs):
```
Comment (forman) [inline xcube_geodb/cli/get_by_bbox.py]: Missing description. What does this command do?

File: xcube_geodb/cli/get_by_bbox.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# The MIT License (MIT)
+# Copyright (c) 2019 by the xcube development team and contributors
```
Comment (forman) [inline xcube_geodb/cli/get_by_bbox.py]:
```suggestion
# Copyright (c) 2020 by the xcube development team and contributors
```

File: xcube_geodb/cli/main.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# The MIT License (MIT)
+# Copyright (c) 2019 by the xcube development team and contributors
```
Comment (forman) [inline xcube_geodb/cli/main.py]:
```suggestion
# Copyright (c) 2020 by the xcube development team and contributors
```

File: xcube_geodb/auth0/client_grant.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+  "scope": [
+    "user:helge"
+  ]
+}
```
Comment (forman) [inline xcube_geodb/auth0/client_grant.json]: What is the use of this file?

File: xcube_geodb/auth0/user.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  "given_name": "John",
+  "family_name": "Doe",
+  "name": "John Doe",
+  "nickname": "Johnny",
+  "picture": "https://secure.gravatar.com/avatar/15626c5e0c749cb912f9d1ad48dba440?s=480&r=pg&d=https%3A%2F%2Fssl.gstatic.com%2Fs2%2Fprofiles%2Fimages%2Fsilhouette80.png",
+  "connection": "Username-Password-Authentication",
+  "password": "youpasswordhash",
+  "verify_email": false
+}
+
```
Comment (forman) [inline xcube_geodb/auth0/user.json]: What is the use of this file?

File: xcube_geodb/auth0/client.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  "jwt_configuration": {
+    "lifetime_in_seconds": 36000,
+    "scopes": {"user:dochelge":  "Access to user schema", "schema:write":  "Write access to schema"},
+    "alg": "RS256"
+  },
+  "sso": false,
+  "sso_disabled": false,
+  "custom_login_page_on": false,
+  "client_metadata": {}
+}
```
Comment (forman) [inline xcube_geodb/auth0/client.json]: What is the use of this file?

File: xcube_geodb/cli/common.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
+from typing import Dict, Any
+
+import click
+
+
+def new_cli_ctx_obj():
+    return {
+        "traceback": False,
+        "scheduler": None,
```
Comment (forman) [inline xcube_geodb/cli/common.py]: ?

File: xcube_geodb/cli/common.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    return click.option(
+        '--traceback',
+        is_flag=True,
+        help="Enable tracing back errors by dumping the Python call stack. "
+             "Pass as very first option to also trace back error during command-line validation.",
+        callback=_callback)(func)
+
+
+def parse_cli_kwargs(value: str, metavar: str = None) -> Dict[str, Any]:
```
Comment (forman) [inline xcube_geodb/cli/common.py]: Code duplication with xcube?

File: xcube_geodb/cli/common.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+            if metavar:
+                message = f"Invalid value for {metavar}: {value!r}"
+            else:
+                message = f"Invalid value: {value!r}"
+            raise click.ClickException(message)
+    else:
+        return dict()
+
+
+def handle_cli_exception(e: BaseException, exit_code: int = None, traceback_mode: bool = False) -> int:
```
Comment (forman) [inline xcube_geodb/cli/common.py]: Code duplication with xcube?

File: xcube_geodb/core/collections.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+from typing import Dict
+
+
+class Collections:
+    def __init__(self, config: Dict):
```
Comment (forman) [inline xcube_geodb/core/collections.py]: Pydocs? Not clear, what the purpose is.

File: xcube_geodb/core/collections.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        self._config = config
+
+    @property
+    def config(self):
```
Comment (forman) [inline xcube_geodb/core/collections.py]: Which config? Type annotation?

File: xcube_geodb/core/message.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
+class Message:
+    def __init__(self, message: str):
+        self._message = message
+
+    @property
+    def message(self):
```
Comment (forman) [inline xcube_geodb/core/message.py]: A class named `Message` should not have a property with the same name. Consider `text` or `content`. Add type annotations.

File: xcube_geodb/core/user.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+import requests
+from dotenv import find_dotenv, load_dotenv
+from geoserver import settings
+
+from xcube_geodb.defaults import GEOSERVER_DEFAULT_PARAMETERS
+
+LOGGER = logging.getLogger("gsconfig2.catalog")
+
+
+class GeoserverSession:
```
Comment (forman) [inline xcube_geodb/core/user.py]:
```suggestion
class GeoServerSession:
```

File: xcube_geodb/core/user.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+from abc import abstractmethod
+from typing import Dict, Optional
+from urllib.parse import urljoin
+import logging
+from xml.etree.ElementTree import XML
+import requests
+from dotenv import find_dotenv, load_dotenv
+from geoserver import settings
+
+from xcube_geodb.defaults import GEOSERVER_DEFAULT_PARAMETERS
```
Comment (forman) [inline xcube_geodb/core/user.py]:
```suggestion
from xcube_geodb.defaults import GEO_SERVER_DEFAULT_PARAMETERS
```

File: xcube_geodb/core/user.py [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+         as long as it's a string
+       * otherwise, we raise a ValueError
+    """
+    if isinstance(named, str) or named is None:
+        return named
+    elif hasattr(named, 'name') and isinstance(named.name, str):
+        return named.name
+    else:
+        msg = f"Can't interpret object as a name or a configuration object"
+        raise ValueError(msg.format(named))
```
Comment (forman) [inline xcube_geodb/core/user.py]: `named` not used.

File: xcube_geodb/core/user.py [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+       * the "name" of an object with a property named name is that property -
+         as long as it's a string
+       * otherwise, we raise a ValueError
+    """
+    if isinstance(named, str) or named is None:
+        return named
+    elif hasattr(named, 'name') and isinstance(named.name, str):
+        return named.name
+    else:
+        msg = f"Can't interpret object as a name or a configuration object"
```
Comment (forman) [inline xcube_geodb/core/user.py]:
```suggestion
        msg = "Can't interpret object as a name or a configuration object"
```

File: xcube_geodb/core/user.py [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+
+    def __init__(self,
+                 user_name: str = GEOSERVER_DEFAULT_PARAMETERS['user_name'],
+                 password: str = GEOSERVER_DEFAULT_PARAMETERS['password'],
+                 active: bool = True,
+                 session: Optional[GeoserverSession] = None
+                 ):
+        super().__init__(session=session)
+
+        self._username = user_name
```
Comment (forman) [inline xcube_geodb/core/user.py]: 1. `user_name`
2. `_username`

File: xcube_geodb/core/user.py [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+    @classmethod
+    def load(cls, user_name: str, session: Optional[GeoserverSession] = None):
+        session = session or GeoserverSession()
+
+        r = session.get(urljoin(User.path, user_name))
+        user_json = r.json()
+
+        return cls(
+            user_name=user_json['userName'],
```
Comment (forman) [inline xcube_geodb/core/user.py]: 3. `userName`

--> 3 different forms. Please align naming.

File: xcube_geodb/core/user.py [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+from geoserver import settings
+
+from xcube_geodb.defaults import GEOSERVER_DEFAULT_PARAMETERS
+
+LOGGER = logging.getLogger("gsconfig2.catalog")
+
+
+class GeoserverSession:
+    """
+    The GeoServer catalog represents all of the information in the GeoServer
```
Comment (forman) [inline xcube_geodb/core/user.py]: Class is named `GeoserverSession` but docs say "GeoServer catalog". Please clarify.
Also explain the purpose of the class.

File: xcube_geodb/core/user.py (hunk 14 above)
Comment (forman) [reply xcube_geodb/core/user.py]: The GeoServer project explicitly uses camel-case writing of their name.

File: xcube_geodb/core/user.py (hunk 20 above)
Comment (forman) [reply xcube_geodb/core/user.py]: Why is it in a module called `user`?

File: xcube_geodb/core/user.py [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+    if isinstance(named, str) or named is None:
+        return named
+    elif hasattr(named, 'name') and isinstance(named.name, str):
+        return named.name
+    else:
+        msg = f"Can't interpret object as a name or a configuration object"
+        raise ValueError(msg.format(named))
+
+
+class UploadError(Exception):
```
Comment (forman) [inline xcube_geodb/core/user.py]: Place in module `errors`?

File: xcube_geodb/core/user.py (hunk 18 above)
Comment (dzelge) [reply xcube_geodb/core/user.py]: The user is not used at this stage. I have removed it from this version but applied your changes for the next version where we are going to introduce a use class.

File: xcube_geodb/core/user.py (hunk 20 above)
Comment (dzelge) [reply xcube_geodb/core/user.py]: Same as for class user.

**Pull request merged**